projection/validator_delegation/view/redelegations.go

```diff
@@ -27,18 +37,166 @@ func NewRedelegationsView(handle *rdb.Handle) Redelegations {
+
+	sql, sqlArgs, err := view.rdb.StmtBuilder.
+		Insert(
+			"view_vd_redelegations",
+		).
+		Columns(
+			"height",
+			"delegator_address",
+			"validator_src_address",
+			"validator_dst_address",
+			"entries",
+		).
+		Values(
+			fmt.Sprintf("[%v,)", row.Height),
+			row.DelegatorAddress,
+			row.ValidatorSrcAddress,
+			row.ValidatorDstAddress,
+			entriesJSON,
+		).
+		ToSql()
+
+	if err != nil {
+		return fmt.Errorf("error building redelegation insertion sql: %v: %w", err, rdb.ErrBuildSQLStmt)
+	}
+
+	result, err := view.rdb.Exec(sql, sqlArgs...)
+	if err != nil {
+		return fmt.Errorf("error inserting redelegation into the table: %v: %w", err, rdb.ErrWrite)
+	}
+	if result.RowsAffected() != 1 {
+		return fmt.Errorf("error inserting redelegation into the table: no row upserted: %w", rdb.ErrWrite)
+	}
 
 	return nil
 }
 
-func (view *RedelegationsView) Upsert(row RedelegationRow) error {
+func (view *RedelegationsView) Update(row RedelegationRow) error {
+
+	// Check if there is an record's lower bound start with this height.
+	found, err := view.checkIfRedelegationRecordExistByHeightLowerBound(row)
+	if err != nil {
+		return fmt.Errorf("error in checking new Redelegation record existence at this height: %v", err)
+	}
+
+	if found {
+		// If there is a record that height = `[row.Height,)`, then update the existed one
+
+		entriesJSON, err := jsoniter.MarshalToString(row.Entries)
+		if err != nil {
+			return fmt.Errorf("error JSON marshalling RedelegationRow.Entries for upsertion: %v: %w", err, rdb.ErrBuildSQLStmt)
+		}
+
+		sql, sqlArgs, err := view.rdb.StmtBuilder.
+			Update(
+				"view_vd_redelegations",
+			).
+			SetMap(map[string]interface{}{
+				"entries": entriesJSON,
+			}).
+			Where(
+				"delegator_address = ? AND validator_src_address = ? AND validator_dst_address = ? AND height @> ?::int8",
+				row.DelegatorAddress,
+				row.ValidatorSrcAddress,
+				row.ValidatorDstAddress,
+				row.Height,
+			).
+			ToSql()
+
+		if err != nil {
+			return fmt.Errorf("error building Redelegation update sql: %v: %w", err, rdb.ErrBuildSQLStmt)
+		}
+
+		result, err := view.rdb.Exec(sql, sqlArgs...)
+		if err != nil {
+			return fmt.Errorf("error updating Redelegation into the table: %v: %w", err, rdb.ErrWrite)
+		}
+		if result.RowsAffected() != 1 {
+			return fmt.Errorf("error updating Redelegation into the table: row updated: %v: %w", result.RowsAffected(), rdb.ErrWrite)
+		}
+
+	} else {
+		// If there is not an existed record, then update the previous record's height range and insert a new record
+
+		err := view.setRedelegationRecordHeightUpperBound(row)
+		if err != nil {
+			return fmt.Errorf("error updating Redelegation.Height upper bound: %v", err)
+		}
+		err = view.Insert(row)
+		if err != nil {
+			return fmt.Errorf("error inserting a new record for this Redelegation: %v", err)
+		}
+
+	}
 
 	return nil
 }
 
+func (view *RedelegationsView) checkIfRedelegationRecordExistByHeightLowerBound(row RedelegationRow) (bool, error) {
```

Comment:
suggest change to`IsExistedByLowerBoundHeight(delegatorAddress, validatorSrcAddress, validatorDstAddress string, height int64)` for readability

Reply:
Thanks, updated.